doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -250,87 +307,235 @@ an ``__array_function__`` method that is identical to
-
-We like the idea of writing overloads with minimal syntax, but dynamic
-code generation also has potential downsides, such as slower import times, less
-transparent code and added difficulty for static analysis tools. It's not clear
-that tradeoffs would be worth it, especially because functions with complex
-signatures like ``np.einsum`` would assuredly still need to invoke
-``try_array_function_override`` directly.
-
-So we don't propose adding such a decorator yet, but it's something worth
-considering for the future.
+    def sum(array, ..., keepdims=np._NoValue):
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+        return array.sum(..., **kwargs)
+
+This also makes it possible to add optional arguments to ``__array_function__``
+implementations incrementally and only in cases where it makes sense. For
+example, a library implementing immutable arrays would not be required to
+explicitly include an unsupported ``out`` argument. Doing this properly for all
+optional arguments is somewhat onerous, e.g.,
+
+.. code:: python
+
+    def my_sum(array, ..., out=None):
+        if out is not None:
+            raise TypeError('out argument is not supported')
+        ...
+
+We thus avoid encouraging the tempting shortcut of adding catch-all
+``**ignored_kwargs`` to the signatures of functions called by NumPy, which fails
+silently for mispelled or ignored arguments.
+
+Outside of NumPy
+''''''''''''''''
+
+We also allow ``__array_function__`` implementations to add their own
+optional keyword arguments by including ``**ignored_kwargs`` in dispatcher
+functions. Such implementation-specific arguments are quite common
+(e.g., ``dask.array.sum()`` adds ``split_every`` and ``tensorflow.reduce_sum()``
+adds ``name``), and we believe that adding support for them will add valuable
+flexibility for using NumPy as a high level API.
+
+This will be particularly useful for libraries that implement new high-level
+array functions on top of NumPy functions, e.g.,
+
+.. code:: python
+
+    def mean_squared_error(x, y, **kwargs):
+        return np.mean((x - y) ** 2, **kwargs)
+
+Otherwise, we would need separate versions of ``mean_squared_error`` for each
+array implementation in order to pass specific arguments.
+
+We don't allow adding optional positional arguments, because these are
+reserved for future use by NumPy itself, but conflicts between keyword arguments
+should be relatively rare.
+
+This flexibility does come with costs. In particular, it implicitly adds
+``**kwargs`` to the signature for all wrapped NumPy functions without actually
+including it (because we use ``functools.wraps``). This means it is unlikely
+to work well with static analysis tools, which could report invalid arguments.
+Likewise, there is a price in readability: these optional arguments won't be
+included in the docstrings for NumPy functions. These downsides mean that users
+should make use of this feature with care.
```

Comment:
I wrote this section to justify passing on unrecognized `**kwargs`, but now that I've considered the downsides I no longer think it's a good idea! I will probably move this to the section on "Alternatives".

Reply:
I actually agree with you. If someone needs implementation-specific kwargs, they can just call the library's implementation directly, instead of NumPy's.

Reply:
Also agreed - better to keep it simple.

Reply:
Done